TEMPLATE/package/install.sls

```diff
@@ -3,8 +3,18 @@
 
 {#- Get the `tplroot` from `tpldir` #}
 {%- set tplroot = tpldir.split('/')[0] %}
+{%- set sls_repo = tplroot ~ '.repo.install' %}
 {%- from tplroot ~ "/map.jinja" import TEMPLATE with context %}
 
+         {%- if TEMPLATE.pkg.use_upstream_repo %}
+include:
+  - {{ sls_repo }}
+         {%- endif %}
+
 TEMPLATE-package-install-pkg-installed:
   pkg.installed:
     - name: {{ TEMPLATE.pkg.name }}
+         {%- if TEMPLATE.pkg.use_upstream_repo %}
+    - require:
+      - sls: {{ sls_repo }}
+         {%- endif %}
```

Comment:
Why not building a `require` jinja variable at include time?

Maybe like:

```sls
{%-   set requires = [] %}
{%- if TEMPLATE.pkg.use_upstream_repo %}
{%-   set requires.append({'sls': sls_repo}) %}
include:
  - {{ sls_repo }}
{%- endif %}

TEMPLATE-package-install-pkg-installed:
  pkg.installed:
    - name: {{ TEMPLATE.pkg.name }}
    - require: {{ requires }}
```

I think we shoud try to avoid conditionnal as much as possible in the middle of a state as [explained in conventions](https://docs.saltstack.com/en/latest/topics/development/conventions/formulas.html#light-conditionals-and-looping).

Reply:
I like this suggestion @baby-gnu It's both pythonic and jinjaic ;-)
Yes, I don't like including too much Jinja logic in SLS states.

Reply:
However, I'm not sure [conditionals-and-looping ](https://docs.saltstack.com/en/latest/topics/development/conventions/formulas.html#light-conditionals-and-looping ) applies to this use case.

The install.sls looks okay, just one include.
```
{%- set requires = [] %}
{%- if TEMPLATE.pkg.use_upstream_repo %}
       {%-   set requires.append({'sls': sls_repo}) %}
include:
  - {{ sls_repo }}
{%- endif %}
```
But the clean.sls becomes messy - multiple includes
```
{%- set requires = [] %}
{%- set requires.append({'sls': sls_config_clean}) %}
include:
  - {{ sls_config_clean }}
{%- if TEMPLATE.pkg.use_upstream_repo %}
       {%-   set requires.append({'sls': sls_repo_clean}) %}
  - {{ sls_repo_clean }}
{%- endif %}
```

Reply:
Some kind of syntax issue?
```
           Rendering SLS 'base:TEMPLATE.package.install' failed: Jinja syntax error: expected token 'end of statement block', got '('; line 10

       {%-   set requires.append({'sls': sls_repo}) %}    <======================
```

Reply:
I did not test it ;-)

ho yes, the problem is `set` it should be a `do`:

```
{%-    do requires.append({'sls': sls_repo}) %}
```

I think the same idea could be done for the include in `clean.sls`:

``` sls
{%- set includes = [sls_config_clean] %}
{%- set requires = [{'sls': sls_config_clean}] %}
{%- if TEMPLATE.pkg.use_upstream_repo %}
       {%-   do includes.append(sls_repo_clean) %}
       {%-   do requires.append({'sls': sls_repo_clean}) %}
{%- endif %}

include: {{ includes }}
```

It could be possible to deduce `include` from `requires` using jinja [map](https://jinja.palletsprojects.com/en/2.11.x/templates/#map), the following is untested

``` sls
{%- set requires = [{'sls': sls_config_clean}] %}
{%- if TEMPLATE.pkg.use_upstream_repo %}
       {%-   do requires.append({'sls': sls_repo_clean}) %}
{%- endif %}

include: {{ requires | map(attribute='sls') }}
```

Regards.

Reply:
> It could be possible to deduce `include` from `requires` using jinja [map](https://jinja.palletsprojects.com/en/2.11.x/templates/#map), the following is untested

After some tests I found 2 issues:
1. `map` return an iterator which could not be used directly in `include`, we need the `list` filter
2. it may have some encoding error with Python2

Here is a working example:

```sls
{#- Get the `tplroot` from `tpldir` #}
{%- set tplroot = tpldir.split("/")[0] %}
{%- set sls_repo = tplroot ~ ".repo.install" %}
{%- from tplroot ~ "/map.jinja" import TEMPLATE with context %}


{%- if TEMPLATE.pkg.use_upstream_repo %}
{%-   set requires = [{"sls": sls_repo}] %}
{%- else %}
{%-   set requires = [] %}
{%- endif %}

{#- serialize as YAML to avoid encoding errors on Python2 #}
include: {{ requires | map(attribute="sls") | list | yaml }}

TEMPLATE-package-install-pkg-installed:
  pkg.installed:
    - name: {{ TEMPLATE.pkg.name }}
    - require: {{ requires | yaml }}
```

Reply:
@baby-gnu I reckon our policy should be to completely ignore Python 2 nowadays.  We've already got enough on our plate and it would be great to streamline formulas as much as possible.

Reply:
My problem is that the [salt-bootstrap](https://github.com/saltstack/salt-bootstrap) script installed the python2 version by default on a Bionic VM :-/

Reply:
That should be fixed. Meanwhile, you can pass `-x python3` to the `salt-bootstrap` command line to get that fixed.

Reply:
> That should be fixed. Meanwhile, you can pass `-x python3` to the `salt-bootstrap` command line to get that fixed.

Yes, having some troubles about Jinja which could not find a `mapdata.foo` because the `foo` was `mapdata[u'foo']` I finally thought about the python2 issue and found the `-x python3` option.